File: lib/test/structuredfield/decodeSfDict.test.ts
import assert from 'node:assert';
import test from 'node:test';
import { SfItem } from '../../src/structuredfield/SfItem.js';
import { decodeSfDict } from '../../src/structuredfield/decodeSfDict.js';

test('decodeSfDict', () => {
	assert.deepStrictEqual(decodeSfDict(``), {});
	assert.deepStrictEqual(decodeSfDict(`a=(1 2), b=3, c=4;aa=bb, d=(5 6);valid`), {
		'a': new SfItem([1, 2]),
		'b': new SfItem(3),
		'c': new SfItem(4, { 'aa': Symbol.for('bb') }),
		'd': new SfItem([5, 6], { 'valid': true }),
	});
	assert.throws(() => decodeSfDict(`a=1, b=2)`), (error: any) => {
		assert.deepStrictEqual(error.message, `failed to parse "a=1, b=2)" as Dict`);
		assert.deepStrictEqual(error.cause.message, `failed to parse ")" as Dict`);
		return true;
	});
});

File: python/src/test/structuredfield/decodeSfDict_test.py
import unittest

from ...structuredfield.decodeSfDict import decodeSfDict
from ...structuredfield.SfException import SfException
from ...structuredfield.SfItem import SfItem
from ...structuredfield.SfToken import SfToken


class TestDecodeSfDict(unittest.TestCase):
    def test_blank_string(self):
        self.assertEqual(decodeSfDict(""), {})

    def test_valid_input(self):
        self.assertEqual(
            decodeSfDict("a=(1 2), b=3, c=4;aa=bb, d=(5 6);valid"),
            {
                "a": SfItem([1, 2]),
                "b": SfItem(3),
                "c": SfItem(4, {"aa": SfToken("bb")}),
                "d": SfItem([5, 6], {"valid": True}),
            },
        )

    def test_invalid_input(self):
        with self.assertRaises(SfException) as context:
            decodeSfDict("a=1, b=2)")
        self.assertEqual(
            context.exception.message,
            'failed to parse "a=1, b=2)" as Dict',
        )
        self.assertEqual(
            context.exception.cause.message,
            'failed to parse ")" as Dict',
        )
